### PR TITLE
IAM: Allowed user-defined password for login profiles

### PR DESCRIPTION
**Pull request by Ninir**
#### Reasoning for this change
This is the continuation of https://github.com/hashicorp/terraform/pull/9605.

The previous PR introduced PGP-based keys, encrypting and decrypting using keybase.io.
As the login_profile password was generated, it was missing the ability to generate a custom password. 
This allows to set it, making this even more flexible.

(Migrated from https://github.com/hashicorp/terraform/pull/10734)

#### Real Configuration
```hcl
 resource "aws_iam_user" "user" {
     name = "demo"
     path = "/"
     force_destroy = true
 }
 
 data "aws_caller_identity" "current" {}
 
 data "aws_iam_policy_document" "user" {
     statement {
         effect = "Allow"
         actions = ["iam:GetAccountPasswordPolicy"]
         resources = ["*"]
     }
     statement {
         effect = "Allow"
         actions = ["iam:ChangePassword"]
         resources = ["arn:aws:iam::${data.aws_caller_identity.current.account_id}:user/&{aws:username}"]
     }
 }
 
 resource "aws_iam_user_policy" "user" {
     name = "AllowChangeOwnPassword"
     user = "${aws_iam_user.user.name}"
     policy = "${data.aws_iam_policy_document.user.json}"
 }
 
 resource "aws_iam_access_key" "user" {
     user = "${aws_iam_user.user.name}"
 }
 
 resource "aws_iam_user_login_profile" "user" {
     user = "${aws_iam_user.user.name}"
     password = "test"
 }
```

#### Use cases
Demos, trainings or handsons.

#### Acceptance tests
```
$ make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSUserLoginProfile_'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/12/14 21:58:25 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSUserLoginProfile_ -timeout 120m
=== RUN   TestAccAWSUserLoginProfile_basic
--- PASS: TestAccAWSUserLoginProfile_basic (22.11s)
=== RUN   TestAccAWSUserLoginProfile_keybase
--- PASS: TestAccAWSUserLoginProfile_keybase (20.71s)
=== RUN   TestAccAWSUserLoginProfile_keybaseDoesntExist
--- PASS: TestAccAWSUserLoginProfile_keybaseDoesntExist (12.11s)
=== RUN   TestAccAWSUserLoginProfile_notAKey
--- PASS: TestAccAWSUserLoginProfile_notAKey (14.28s)
=== RUN   TestAccAWSUserLoginProfile_customPassword
--- PASS: TestAccAWSUserLoginProfile_customPassword (19.61s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	88.850s
```

### CODE REVIEW

File: aws/resource_aws_iam_user_login_profile_test.go [hunk 1]
```diff
@@ -169,7 +189,7 @@ func testDecryptPasswordAndTest(nProfile, nAccessKey, key string) resource.TestC
 				NewPassword: aws.String(generatePassword(20)),
 			})
 			if err != nil {
-				if awserr, ok := err.(awserr.Error); ok && awserr.Code() == "InvalidClientTokenId" {
+				if awserr, ok := err.(awserr.Error); ok && (awserr.Code() == "InvalidClientTokenId" || awserr.Code() == "EntityTemporarilyUnmodifiable") {
```
Comment (Ninir) [inline aws/resource_aws_iam_user_login_profile_test.go]: This is required, otherwise this happens

> Check failed: Check 2/2 error: Error changing decrypted password: EntityTemporarilyUnmodifiable: Login Profile for User test-user-5423997845776594466 cannot be modified while login profile is being created.

**Comment by Ninir** [comment]
If you try to change the password without impersonating the user, the following error will be thrown:

    InvalidUserType: Only IAM Users can change their own password.

I am able to do the update stuff, but it will require to impersonate the user, and update the profile.
Do you think we should go this way, or prefer the use of ForceNew on the password?

**Comment by Ninir** [comment]
@radeksimko Would it be possible for you to review this one please? 😄 

'Managed to handle the update using the UpdateLoginProfile. I was using the ChangePassword call previously, which requires to be impersonated as the user.

File: aws/resource_aws_iam_user_login_profile.go [hunk 2]
```diff
@@ -19,30 +19,38 @@ func resourceAwsIamUserLoginProfile() *schema.Resource {
 	return &schema.Resource{
 		Create: resourceAwsIamUserLoginProfileCreate,
 		Read:   schema.Noop,
-		Update: schema.Noop,
+		Update: resourceAwsIamUserLoginProfileUpdate,
 		Delete: schema.RemoveFromState,
```
Comment (Ninir) [inline aws/resource_aws_iam_user_login_profile.go]: I think we can also handle the Delete in another request: if you try to create a login profile for a user that has already been created, you will get a ConflictException. When we don't want to keep the login profile, we don't want only to remove it from the state: we actually want to delete it, to avoid conflicts.

**Comment by radeksimko** [comment]
I appreciate the time you invested into this, but I'm not too strongly inclined to the idea of giving folks the option of specifying custom password, because (a) the story around secrets in state isn't great and (b) we should be promoting good practices when dealing with sensitive data generally and shared passwords don't belong into good practices.

I'm genuinely worried people are going to use this instead of letting AWS generate the password and they will end up in secrets sprawl unknowingly or just because "GPG seemed to hard to grok".

@catsby What do you think?

**Comment by catsby** [comment]
I agree with Radek here. I'm grateful for the time and energy you've put into this, however I feel like we need to encourage best practices with regards to security here and have users go the extra step(s) for pgp. 

That said, I'd like to close this issue for now. Please let me know what you think.

**Comment by Ninir** [comment]
Using PGP is certainly a good security practice. This work was started a long time ago, as we often need to create on-the-fly users, with following passwords (team1, team2, team3, etc.), whether it is for trainings / "hands-on" with clients.

As far as I understand, Keybase.io, for instance, could help me doing this: encrypt/decrypt passwords and return them all as outputs.
I would then need to find a way to provide passwords to clients / attendees, like online Google doc / spreadsheet. Am I correct?

What about being able to set your password BUT encrypt it using PGP? would use best of both worlds IMO.

Thank you for taking the time to talk about this (available on Slack for any longer discussion 😄 )!

**Comment by radeksimko** [comment]
> What about being able to set your password BUT encrypt it using PGP?

I think that's defeating the purpose of the encryption as you (or anyone else putting it into the config) practically know what the password is, unless you're also generating it using data sources from the `random` provider (in which case you can as well just let AWS do it) - and even then the random string will be visible in its raw form in plan, apply output and in the state file.

For your use case of trainings - is there any particular reason you can't use a single GPG/PGP key and decrypt it for them? I'm not sure what your exact workflow is. e.g. do _you_ `apply` these configs for them prior to the training or do they do it? - presumably it's you as I doubt you'd want to give every trainee the privilege of creating an IAM user 😃 . So you could decrypt it for them as part of the preparation using the same (your own) key for all passwords.

TL;DR why not add 1 extra step of `terraform console/output ... | gpg -d`?

Eventually if you want to do it the _right way_ ™️  you could let every trainee create a GPG key as part of the training and let them decrypt it.

> available on Slack for any longer discussion

yep - I'm also available there.

Do you mind me closing this issue now?

**Comment by Ninir** [comment]
> TL;DR why not add 1 extra step of terraform console/output ... | gpg -d?

> For your use case of trainings - is there any particular reason you can't use a single GPG/PGP key and decrypt it for them? I'm not sure what your exact workflow is. e.g. do you apply these configs for them prior to the training or do they do it? - presumably it's you as I doubt you'd want to give every trainee the privilege of creating an IAM user 😃 . So you could decrypt it for them as part of the preparation using the same (your own) key for all passwords.

Yup, that was my point when writing: `As far as I understand, Keybase.io, for instance, could help me doing this: encrypt/decrypt passwords and return them all as outputs.`!

Regarding your other point:
> I think that's defeating the purpose of the encryption as you (or anyone else putting it into the config) practically know what the password is, unless you're also generating it using data sources from the random provider (in which case you can as well just let AWS do it) - and even then the random string will be visible in its raw form in plan, apply output and in the state file.

Having to generate a password is fine for me, as soon as the overall process is fine: generating, encrypting/decrypting using a **SINGLE** key, and being able to provide passwords as outputs, so that I can share passwords for each trainee.
I'm not against changing the process as long as it can take as much time, or if it does not bring a lot of issues: the more tools & steps you have, the more difficult it is to set-up...

> Eventually if you want to do it the right way ™️ you could let every trainee create a GPG key as part of the training and let them decrypt it.

Agreed, however that is not an option when you have non-technical people: time being lost creating your GPG key, managing your env, and so on... we can't afford that.

If you can ensure me that my main point is fine (having a single key & managing the whole process) then feel free to close this one 😄 
Otherwise... 🤷‍♂️ 😄

**Comment by radeksimko** [comment]
> If you can ensure me that my main point is fine (having a single key & managing the whole process) then feel free to close this one 😄

Yes, and you don't even need keybase for this. Keybase is just a place to store your pubkey so you can refer to it. The Keybase CLI tools which wrap GPG make the process slightly easier, but if you're already familiar with GPG/PGP you can use [GPG Tools](https://gpgtools.org/) to achieve this and then just enter base-64 encoded pubkey (`cat ~/my-pubkey.asc | base64`) into the `pgp_key` field per docs: https://www.terraform.io/docs/providers/aws/r/iam_user_login_profile.html#pgp_key

Feel free to ping me if you need help with the above.
